modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/delegate/delegate.es.js

```diff
@@ -45,6 +45,8 @@ function delegate(element, eventName, selector, callback) {
 			!defaultPrevented &&
 			(target.matches(selector) || target.closest(selector))
 		) {
+			event.delegateTarget = target.closest(selector);
```

Comment:
Looks good, seeing the other condition above makes me a bit uneasy, though.

@wincent, @bryceosterhaus, @kresimir-coko, why do we have the `target.matches(selector) || target.closest(selector)` above?

`target.matches(selector)` would return `target` when `target.matches(selector)` is `true`. Isn't that a redundant truthy condition? I think I remember some comments about this but I'm not sure what were they about.

Reply:
My understanding is this: `target.matches(selector)` is a bit like an identity check ("is this thing the thing I am looking for?") whereas `target.closest(selector)` is an ancestry check ("where is the closest thing?") — but if `target` _is_ the thing both will return `true`.

So you might think that you could just drop the `target.matches()` because `target.closest()` will always return `true` whenever `target.matches()` does, therefore covering both — but there is a slim argument to be made that it is a short-circuiting performance optimization:

- First try cheap check (`matches()`).
- Only if that fails try expensive check (`closest()`).

In practice, I don't think the argument is very strong, because the "expensive" check should be quick when the target is the same element, and when the quick check fails, you are going to do the expensive check anyway (potentially walking up the entire DOM).

Reply:
> In practice, I don't think the argument is very strong, because the "expensive" check should be quick when the target is the same element, and when the quick check fails, you are going to do the expensive check anyway (potentially walking up the entire DOM).

Yeah, this was the point I wanted to make if this happened to be an attempt at shortcutting the conditionals. `closest` should always be quick when `target.matches` holds true, so we might paying for it with more unnecessary checks for `matches` that make it slower on average.

Reply:
This is not blocking anyways, if green, let's get this forward!!

Reply:
I missed @wincent's [comment about the tests](https://github.com/liferay-frontend/liferay-portal/pull/531#issuecomment-732793923), so that should of course be addressed ;)

Reply:
> Yeah, this was the point I wanted to make if this happened to be an attempt at shortcutting the conditionals. `closest` should always be quick when `target.matches` holds true, so we might paying for it with more unnecessary checks for `matches` that make it slower on average.

About the only other reason I can think of is that, conceptually, having the `||` expression makes it clear that there are two cases that we care about: clicking directly, and clicking somewhere inside — it's true that the first one is subsumed by the second, but having both calls out that we _thought_ about both cases.

(I'm not actually arguing that we keep the expression; just trying to figure out why we might have it.)

Reply:
I'm the one that added the `|| target.closest(selector)`, now that you put it like that, yea, it sounds like `target.closest(selector)` is enough in that condition. Good catch @jbalsas !

I checked their specs and `.matches` returns a `Boolean`, whereas `.closest` returns an `Element`, but in this case `.closest` does everything that we need without the use of `.matches` which makes `.matches` a redundant truthy condition.

Reply:
One final thought... seeing as we are micro-optimizing to get rid of a superfluous `matches()` call, maybe we should avoid the repeated `closest()` calls too.

ie. instead of:

```js
if (eventName === 'click' && isDisabled(target)) {
	return;
}

if (!defaultPrevented && target.closest(selector)) {
	event.delegateTarget = target.closest(selector);

	callback(event);
}
```

this:

```js
if (defaultPrevented || (eventName === "click" && isDisabled(target))) {
	return;
}

const delegateTarget = target.closest(selector);

if (delegateTarget) {
	event.delegateTarget = delegateTarget;

	callback(event);
}
```

(Note also hoisting the `defaultPrevent` check up higher to potentially short-circuit and avoid another possibly expensive call to `closest()` instide the `isDisabled()` function.)

Up to you. 🤷  Like @jbalsas said, this looks forwardable.